src/main/java/emailvalidator4j/parser/DomainPart.java

```diff
@@ -112,13 +113,25 @@ private boolean isDomainLiteral(boolean openBrackets) throws UnclosedDomainLiter
         return this.lexer.getCurrent().equals(Tokens.OPENBRACKET);
     }
 
-    private void checkNotAllowedChars(TokenInterface token) throws DomainNotAllowedCharacter {
+    private void checkNotAllowedChars(TokenInterface token) throws DomainNotAllowedCharacter, DomainHyphen {
         if (notAllowedTokens.contains(token)) {
             throw new DomainNotAllowedCharacter(
                     String.format("%s is not allowed in domain part", token.getName())
             );
         }
-
+        if (Tokens.GENERIC.equals(token.getName())) {
+            String subDomain = token.getText();
+            if (subDomain.charAt(0) == '-') {
+                throw new DomainHyphen("Sub-domains cannot start with hyphen");
+            }
+            if (subDomain.charAt(Math.max(0, subDomain.length() - 1)) == '-') {
+                throw new DomainHyphen("Sub-domains cannot end with hyphen");
+            }
+            Matcher matcher = subDomainIllegalCharacters.matcher(subDomain);
+            if (matcher.find()) {
```

Comment:
Search for _any_ character that is not allowed in a domain name. One instance is fatal.